### DISCUSSION TITLE
how to set domainStrategy for a SOCKS outbound?

### DISCUSSION THREAD

**Discussion by pulsarice** [XTLS/Xray-core#3495] [Q&A]
Hi
I have multiple outbounds on my server. One of them is a SOCKS outbound (connecting to another service running on localhost)
I have no control on how that SOCKS outbound resolves domain names. (If the request is sent as domain name, it uses IPv6 by default, unless the request is sent as IPv4)

I want to force that SOCKS outbound to use IPv4 domainStrategy.
I wasn't successful assigning a domainStrategy to that SOCKS outbound.

I managed to achieve it by chaining a freedom outbound to SOCKS.
**Question: Is there a better way to force domainStrategy on a SOCKS outbound?**

```
  "outbounds": [
    {
      ...other outbounds,
    {
      "tag": "socks-ipv4",
      "protocol": "freedom",
      "settings": {
        "domainStrategy": "UseIPv4"
      },
      "proxySettings": {
        "tag": "socks-out"
      }
    },
    {
      "tag": "socks-out",
      "protocol": "socks",
      "settings": {
        "servers": [
          {
            "address": "127.0.0.1",
            "port": 1080
          }
        ]
      }
    },
    ...other outbounds
  ]
```

**Comment by mmmray** [comment]
https://xtls.github.io/en/config/transport.html#streamsettingsobject

should be theoretically doable like this

```
{
  "tag": "socks-out",
  "protocol": "socks",
  "streamSettings": {
    "sockopt": {
      "domainStrategy": "UseIPv4"
    }
  }
}
``` 

untested

**Reply by pulsarice** [reply]
Thank you for your response. I've already tested this, but it still used IPv6!

**Comment by uuonda** [comment]
Forcing IPv4 through socks requires resolving a domain to IPv4 address and then handling it to socks. Have you tried configuring built-in DNS server?

>In the `sockopt` setting, if `domainStrategy` is set to `UseIP`, system connections initiated through the outbound proxy will first be resolved to an IP address using the built-in DNS server before making the connection.

**Reply by pulsarice** [reply]
Yes I have the DNS configured. I think the sockopt domainStrategy actually affects how the outbound "server address" is resolved and does nothing to the requests that are sent via the outbound. (already tested sockopt domainStrategy and was unsuccessful)

**Reply by uuonda** [reply]
@pulsarice Are you forcing IPv4 in your DNS config?

**Reply by pulsarice** [reply]
```
  "dns": {
    "tag": "dns-inbound",
    "servers": [
      "8.8.8.8",
      "1.1.1.1"
    ],
    "queryStrategy": "UseIPv4"
  }
```
I also route user traffic with dest port=53 to internal DNS server.
However because of Sniffing on user side and server side, the requests will be converted to domain name anyway and finally the server will decide how to forward it. I have 3 outbound foreign servers and 1 local SOCKS outbound. Foreign servers will do their own name resolution (using IPv4), but the local SOCKS server will use IPv6 if the request is sent as domain name.

**Reply by uuonda** [reply]
>because of Sniffing on user side and server side, the requests will be converted to domain name anyway

`routeOnly`?
